Improvements to spider_plus

godylockz opened the pull request:
[Fixes regarding downloading files in spider_plus, typos, consistency, linting, module description, added smart "already downloaded" logic. f strings, human size printouts.
Did all the TODOs...

This commit will fix the issue: SMB Shares - Download Files [Bug] #54 
 >> This will remove the file that was downloaded if it is zero size and should not be.

NeffIsBack commented:
Hi thanks for the contribution.
Which line is responsible for fixing #54? The functionality to download files is at https://github.com/mpgn/CrackMapExec/blob/35d2619cfb812dd08bffc4597cf6f2278762bc6d/cme/protocols/smb.py#L1380C9-L1380C17
As far as i can see this commit removes files which are broken when downloaded by the spider_plus module, but not with the normal cme `--get-file` functionality

NeffIsBack commented on cme/modules/spider_plus.py:
```diff
@@ -43,28 +43,29 @@ def make_dirs(path):
 
 
-get_list_from_option = lambda opt: list(map(lambda o: o.lower(), filter(bool, opt.split(","))))
+def get_list_from_option(opt): return list(
+    map(lambda o: o.lower(), filter(bool, opt.split(","))))
```
Please do the line break after `...(opt):`

godylockz replied:
Fixed.

NeffIsBack commented on cme/modules/spider_plus.py:
```diff
@@ -73,11 +74,11 @@ def __init__(
-            self.conn_retry -= 1
-            self.logger.display(f"Reconnect to server {self.conn_retry}")
+        for i in range(1, self.max_connection_attempts+1):
+            self.logger.display(
+                f"Reconnection attempt #{i}/{self.max_connection_attempts} to server.")
```
There are a lot of line breaks in the file where it is not necessary in my oppinion. For ease of read please don't line break only because of line-length (there might be valid reasons if it is easier to read, but not in this case imo)

NeffIsBack replied:
Hmm i am on the other side of that argument haha. In my oppinion the days of a 80 characters screen length are long gone. If the line is too long for you you can simply enable line wrapping in your editor (but screens are huge nowadays). Having these weird line breaks are really confusing to read imo.

Yes i also use autopep8, i just disabled the line-length check. I hate when it does line breaks on its own, rest is nice

godylockz commented:
> Hi thanks for the contribution. Which line is responsible for fixing #54? The functionality to download files is at https://github.com/mpgn/CrackMapExec/blob/35d2619cfb812dd08bffc4597cf6f2278762bc6d/cme/protocols/smb.py#L1380C9-L1380C17 As far as i can see this commit removes files which are broken when downloaded by the spider_plus module, but not with the normal cme `--get-file` functionality

Just added:
- Return Code logic in cme logger. Return code exits in main.
- Added same checking of downloaded file size, added condition to only check when errored as the file size on the share could actually be an empty file.

godylockz commented:
Normal:
![image](https://github.com/mpgn/CrackMapExec/assets/81207744/3ae5c8d0-ef7c-46ae-8117-6e42bff589d9)

Verbose:
![image](https://github.com/mpgn/CrackMapExec/assets/81207744/b245d72c-19ef-4671-a36b-3c267384eba5)

NeffIsBack commented on cme/modules/spider_plus.py:
```diff
@@ -170,169 +195,202 @@ def spider(self):
-                    "mtime_epoch": time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(result.get_mtime_epoch())),
-                    #'atime': time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(result.get_atime())),
-                    "atime_epoch": time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(result.get_atime_epoch())),
+                    "size": human_file_size,
+                    "ctime_epoch": time.strftime(
+                        "%Y-%m-%d %H:%M:%S", time.localtime(file_creation_time)
+                    ),
+                    "mtime_epoch": time.strftime(
+                        "%Y-%m-%d %H:%M:%S", time.localtime(file_modified_time)
+                    ),
+                    "atime_epoch": time.strftime(
+                        "%Y-%m-%d %H:%M:%S", time.localtime(file_access_time)
+                    ),
```
That is for example the best example. The autoformatting broke this all up into 10 lines where there were 4 originally.
Now for me i have a really hard time figguring out which bracket belongs to which. Before i instantly see what is replaced and how the commands work together.

But thats just my oppinion

godylockz replied:
Understood. Will fix these back to normal!

NeffIsBack commented on cme/protocols/smb.py:
```diff
@@ -1378,16 +1378,24 @@ def put_file(self):
+        transfer_failure = False
+        with open(download_path, "wb+") as file:
             try:
-                self.conn.getFile(self.args.share, self.args.get_file[0], file.write)
-                self.logger.success(f"File {self.args.get_file[0]} was transferred to {file_handle}")
+                self.conn.getFile(share_name, remote_path, file.write)
+                self.logger.success(f'File "{remote_path}" was downloaded to "{download_path}"')
             except Exception as e:
-                self.logger.fail(f"Error reading file {self.args.share}: {e}")
+                self.logger.fail(f'Error reading file "{remote_path}" in share "{share_name}": {e}')
+                transfer_failure = True
+                
+        # Check if the file is empty on failure.
+        if transfer_failure and os.path.getsize(download_path) == 0:
+            os.remove(download_path)
```
Just curious. Wouldn't it be easier to wrap the `with` statement with try and except and then do the transfer_failure check in the exception handling part? E.g.:

```python
try:
    with open(doenload_path...
        self.conn...
except Exception as e:
    self.logger.fail
    if os.path.getsize(download_path) == 0:
        os.remove(download_path)
```

NeffIsBack replied:
Might miss something tho. Dunno if with statements and try-except play nice together

godylockz replied:
The with should keep the file handle open until it finishes by failure or nt such that it moves on outside the with. I need the handle to be closed before I can check the file contents and then delete it only if there was a failure. 

I think it's safer to do it outside the with statement.

NeffIsBack replied:
An exception in the with statement should close the file. I think it is best practice to wrap the with statement in try except to handle errors like these.
I am not sure why this was the other way round in the first place but this would be a good point to change it

See: https://stackoverflow.com/a/31766737

godylockz replied:
Fixed, sorry was wrong!

godylockz commented:
I got some fixes in a bit 👍 don't merge yet.

godylockz commented:
All set. Figured i would add some statistics ...
Hope its not too much:

![image](https://github.com/mpgn/CrackMapExec/assets/81207744/f87a2211-9509-4bcf-9525-1fe3ad03b574)

I'm investigating why this doesn't turn it off though ...
`cme smb authority.htb -u 'guest' -p '' -M spider_plus -o STATS_FLAG=False`

godylockz commented:
> Figured i would add some statistics ... Hope its not too much:
> 
> ![image](https://user-images.githubusercontent.com/81207744/254987056-f87a2211-9509-4bcf-9525-1fe3ad03b574.png)
> 
> I'm investigating why this doesn't turn it off though ... `cme smb authority.htb -u 'guest' -p '' -M spider_plus -o STATS_FLAG=False`

Bug in the `moduleloader.py` .. pushing now.

godylockz commented:
Is this fine to incorporate into the module_loader that gets used for all modules?

I'm surprised there isn't something smart already such as a instance-checker based on the default value in the module options.

```python
for option in self.args.module_options:
    key, value = option.split("=", 1)
    if value.lower() == "false":
        value = False
    elif value.lower() == "true":
        value = True
    module_options[str(key).upper()] = value
```

EIther it goes here or in the spider_plus module... this is whats happening ... LOL
```python
>>> bool('False')
True
>>> bool('False')
True
>>> bool('False')
True
>>> bool('FALSE')
True
```

NeffIsBack commented:
> ```python
> >>> bool('False')
> True
> >>> bool('False')
> True
> >>> bool('False')
> True
> >>> bool('FALSE')
> True
> ```

A String is true if it is not empty. It is not the actual text what gets checked.

godylockz commented:
> > ```python
> > >>> bool('False')
> > True
> > >>> bool('False')
> > True
> > >>> bool('False')
> > True
> > >>> bool('FALSE')
> > True
> > ```
> 
> A String is true if it is not empty. It is not the actual text what gets checked.
> 
> But wouldn't be the easiest solution to supply the arg without quotations?

Its not being sent with quotes....
` cme smb authority.htb -u 'guest' -p '' -M spider_plus -o STATS_FLAG=False`

Its because there is nothing checking if its a flag or not. I'm preety sure there is not a way to set a value of false in cme module options right now unless its processed post-string conversion. So for instance if the module gets 'false' then convert it to boolean in the module itself.

ILightThings commented:
The slinky module has an implementation of a bool flag that should be applicable here.
https://github.com/mpgn/CrackMapExec/blob/master/cme/modules/slinky.py#L39

godylockz commented:
I don't really like the current setup, and I suggest we change the default module_options processing for all modules going forward. It's a bit strange that we can't set a simple flag to false in the command line options.

## Module Loader:
https://github.com/mpgn/CrackMapExec/blob/master/cme/loaders/moduleloader.py#L87
Recommend Change - Line 87, `cme/loaders/moduleloader.py`
```python
for option in self.args.module_options:
    key, value = option.split("=", 1)
    if value.lower() == "false":
        value = False
    elif value.lower() == "true":
        value = True
    module_options[str(key).upper()] = value
```

## Slinky
https://github.com/mpgn/CrackMapExec/blob/35d2619cfb812dd08bffc4597cf6f2278762bc6d/cme/modules/slinky.py#L40 This  isn't very clear and will evaluate to True with anything following the `=`. To simplify, you could just use `self.cleanup = True` in Python. I think the only way to set a False flag would be `-o CLEANUP=` without true or false.
```text
bool('')
False
```

## Spider Plus
So, I'm thinking of making these changes, this would simply just find the key in the input to set the value rather than checking the value passed in by the user. This makes it a bit more universal so you could do `DOWNLOAD_FLAG` or `ENABLE_DOWNLOADS`, etc.

Before:
```python
self.download_flag = module_options.get("DOWNLOAD_FLAG", False)
self.stats_flag = module_options.get("STATS_FLAG", True)
```

After:
```python
self.download_flag = False
if any("DOWNLOAD" in key for key in module_options.keys()):
    self.download_flag = True
self.stats_flag = True
if any("STATS" in key for key in module_options.keys()):
    self.stats_flag = False
```

What do you all think? I'd appreciate your input on this.

ILightThings commented:
So if you look at https://github.com/mpgn/CrackMapExec/blob/master/cme/cli.py#L67, you see that the project is using the argparser lib. But due to the fact that it is passing everything after `-o`, there is not a way to do a second argparse parser on the module options because the parser gets the commands directly from the os.args.

Regarding Slinky, I realized it looks a little weird if you don't know the context. Stage 1 of slinky puts a file on every available fileshare. Stage 2 which includes the `CLEANUP=True` tag will remove every previously added file. So you would never need CLEANUP to be enabled by default because you would add and then remove the file in the same action.

If you were to make changes on how modules act going forward, you would also need to make everything backwards compatible or change each module to ensure it still works as needed. Which is a big ask because you would then need to setup an environment to make sure they all still work.

I think adding a little text processing would not be too hard to do. Here is a oneliner that you can do for text processing.
```
Stats_flag = True if staus.lower() == "true" else False
```

Live example
```
status_good = "True"
Stats_flag = True if status_good .lower() == "true" else False #This Line
Stats_flag
>> True

status_bad = "false"
Stats_flag = True if status_bad.lower() == "true" else False
Stats_flag
>> False
```

godylockz commented:
All set here!

mpgn commented:
Tahnks for the PR@godylockz can I start the review or you need to has other things ?

godylockz commented:
> Thanks for the PR @godylockz can I start the review or you need to has other things ?

All set.

mpgn commented:
![image](https://github.com/mpgn/CrackMapExec/assets/5891788/7b19abc2-53c4-4b70-9635-f87de2ee8a67)
![image](https://github.com/mpgn/CrackMapExec/assets/5891788/4ab04444-9a0f-4821-94a8-99fe7f95ef9c)
![image](https://github.com/mpgn/CrackMapExec/assets/5891788/b08e8da5-81ff-420d-90b3-9ad61ec36899)

Nice !

![image](https://github.com/mpgn/CrackMapExec/assets/5891788/aa78fffa-8cc5-4c7f-b788-bf77bb05b7ab)

mpgn commented:
@NeffIsBack I let you check one more time before merging if you want :)

NeffIsBack reviewed (approved):
I already dug through the code a lot and did a static code review. If you didn't run into any problems while running the code i think we are good to go 👍

mpgn commented:
Great, merging, @godylockz contact me on twitter to get your coin :)
